apps/cmp.c

```diff
@@ -0,0 +1,3557 @@
+                            const char *cert_descrip)
+{
+    X509 *x = NULL;
+    BIO *cert = NULL;
+    PW_CB_DATA cb_data;
+
+    cb_data.password = pass;
+    cb_data.prompt_info = file;
+
+    if (format == FORMAT_HTTP) {
+#if !defined(OPENSSL_NO_SOCK)
+        x = X509_load_http(file, NULL, NULL, 10);
+#else
+        CMP_err1("http(s) not supported loading cert from '%s'\n", file);
+#endif
+        goto end;
+    }
+    if (file == NULL) {
+        unbuffer(stdin);
+        cert = dup_bio_in(format);
+    } else {
+        cert = bio_open_default(file, 'r', format);
+    }
+    if (cert == NULL)
+        goto end;
+    if (format == FORMAT_ASN1) {
+        x = d2i_X509_bio(cert, NULL);
+    } else if (format == FORMAT_PEM) {
+        x = PEM_read_bio_X509_AUX(cert, NULL, wrap_password_callback, &cb_data);
+    } else if (format == FORMAT_PKCS12) {
+        EVP_PKEY *pkey = NULL; /* pkey is required for matching the cert */
+
+        if (!load_pkcs12(cert, cert_descrip, wrap_password_callback,
+                         &cb_data, &pkey, &x, NULL))
+            goto end;
+        EVP_PKEY_free(pkey);
+    } else {
+        BIO_printf(bio_err, "bad input format specified for %s file '%s'\n",
+                   cert_descrip, file);
+        goto end;
+    }
+
+ end:
+    if (x == NULL) {
+        ERR_print_errors(bio_err);
+        BIO_printf(bio_err, "unable to load %s from '%s'\n",
+                   cert_descrip, file);
+    }
+    BIO_free(cert);
+    return (x);
+}
+
+/* TODO remove when load_csr() is merged in apps/lib/apps.c (PR #4940) */
+static X509_REQ *load_csr(const char *file, int format, const char *desc)
+{
+    X509_REQ *req = NULL;
+    BIO *in;
+
+    in = bio_open_default(file, 'r', format);
+    if (in == NULL)
+        goto end;
+
+    if (format == FORMAT_ASN1)
+        req = d2i_X509_REQ_bio(in, NULL);
+    else if (format == FORMAT_PEM)
+        req = PEM_read_bio_X509_REQ(in, NULL, NULL, NULL);
+    else if (desc != NULL)
+        BIO_printf(bio_err, "unsupported format for CSR loading\n");
+
+ end:
+    if (req == NULL && desc != NULL)
+        BIO_printf(bio_err, "unable to load X509 request\n");
+    BIO_free(in);
+    return req;
+}
+
+/* TODO potentially move this and related functions to apps/lib/apps.c */
+static int adjust_format(const char **infile, int format, int engine_ok)
+{
+    if (!strncmp(*infile, "http://", 7) || !strncmp(*infile, "https://", 8)) {
+        format = FORMAT_HTTP;
+    } else if (engine_ok && strncmp(*infile, "engine:", 7) == 0) {
+        *infile += 7;
+        format = FORMAT_ENGINE;
+    } else {
+        if (strncmp(*infile, "file:", 5) == 0)
+            *infile += 5;
+        /*
+         * the following is a heuristic whether first to try PEM or DER
+         * or PKCS12 as the input format for files
+         */
+        if (strlen(*infile) >= 4) {
+            const char *extension = *infile + strlen(*infile) - 4;
+
+            if (strncmp(extension, ".crt", 4) == 0
```

Comment:
always strncasecmp() ?

Reply:
I've replaced all `strncmp()` calls by `strncasecmp()` for gracefulness with upper-case input.